### ISSUE TITLE
Not able to reproduce validation set accuracy

### ISSUE THREAD

**Issue by mbcel** [Eromera/erfnet_pytorch#17]
I am currently trying out the ERFNet on the cityscapes dataset. For that I use my own training script but the exact same model implementation as yours. 

The mIoU results that I achieve are at best around 62% mIoU on the Cityscapes validation set for training from scratch. Now I am wondering if I am missing something during training since your validation set results are around 69% mIoU for training from scratch(right?).

What I do is:

- Training Scale: 1024x512 (for testing: bilinear upsampling to 2048x1024)
-  Augmentation: Random translation x/y +-2px; rand. horizontal flipping; input normalization to [-1,1]
- Class balancing with the weights from your script (the class train ids are the same as from the official cityscapes-scripts right? Or did you use a different train id distribution?)
- Learning rate schedule with same lambda function as in your script
- Start learning rate: 5e-4, weight decay: 1e-4
- Batch size: 5 (you used 6 right? But I can't imagine that makes the huge difference)
- Trained for 150 epochs then I picked the best working epoch (epoch 127 in my case) -> 62.3 % mIoU on val. set (did you search for the best epoch or achieved the results with simply the last epoch?)

So do you maybe know if I miss something that could lead to my poor performance? Any help would be appreciated!

**Comment by Eromera** [comment]
Hi! Sorry for the late reply.

It's hard to say where should be the difference. All hyperparameters seem quite similar. Batch size has a great effect but going from 6 to 5 should not make that big difference. Where are you getting the 62.3%? In the output of the main.py during training or using the cityscapes scripts after saving the outputs?

The only thing that sounds problematic is that you mention input normalization to [-1,1]. If you normalize the RGB images during training then you should also normalize them during the evaluation or else the model would not be "used" to those pixel values and this could easily justify the 7% loss in IoU. Are you normalizing when you evaluate in the Val set?

**Comment by mbcel** [comment]
I am getting the 62.3% on the 500 validation set images from the cityscapes dataset. I am setting the model in evaluation mode and I use my own script for evaluation. I also do normalization on the val set images.

How do you normalize the input images? Did you select the model from best working epoch or simply the last epoch?

I will try out your evaluation script and training script and see if I might can find a difference. But thank you for the help!

**Comment by Eromera** [comment]
Hi, 
In the main code normalization is not used but you can normalize with torchvision.transforms.Normalize. 
I normally use the model with best acc in the val set. 

When you train from scratch, are you training encoder and then decoder or only train the full network once without pretraining encoder?

**Comment by mbcel** [comment]
I only train the network once with encoder and decoder together.

**Comment by Eromera** [comment]
Then that should explain the difference. I get 69% by training encoder first in downsampled cityscapes labels (128x64) for some epochs, and then attach the decoder and train the full network again for some epochs. If you train the full network without pretraining encoder, then the encoder filters take much more to "shape" from random init. You could try pretraining encoder or alternatively if you want to train full network in 1 pass then you could a) use an auxiliary loss at encoder's output or b) train for more epochs (maybe 300) or with a different LR schedule (maybe higher LR at the beginning).

**Comment by mbcel** [comment]
Ah okay I gonna try that out and see if that gives me the higher val set accuracy. Thank you very much!

**Comment by marcel1991** [comment]
Okay I am able to reproduce the accuracies that you stated now. 

The highest I achieve from scratch on the validation set is 73.6% mIoU. But the results really vary a lot over different epochs from 50% mIoU up to 73%mIou.

**Comment by mbcel** [comment]
Okay I now do achieve 69% mIoU from scratch. 

The reason for me not being able to achieve your results was the learning rate schedule. I wasn't able to achieve that accuracy with the e-function schedule you used in your code but instead I achieved it with a step lr-schedule similar to the schedule you described in your paper. Reducing lr by 0.5 is thereby important, I used 1/10 first which did not achieve this high performance.

Thank you for your help!
